docat/docat/utils.py

```diff
@@ -168,22 +168,34 @@ def get_project_details(upload_folder_path: Path, project_name: str) -> ProjectD
 
 def index_all_projects(
     upload_folder_path: Path,
-    index_db: TinyDB,
+    index_db_path: Path,
 ):
     """
     This will extract all content from all versions for each project,
     and save it into index.json.
     """
-    # drop existing index
-    index_db.drop_tables()
-    index_db.table("projects")
-    index_db.table("files")
+    TMP_DB_PATH = upload_folder_path / "tmp-index.json"
+
+    # check already indexing
+    if TMP_DB_PATH.exists():
+        return
 
     all_projects = get_all_projects(upload_folder_path).projects
 
-    for project in all_projects:
-        update_version_index_for_project(upload_folder_path, index_db, project.name)
-        update_file_index_for_project(upload_folder_path, index_db, project.name)
+    # make index with seperate db, and swap it afterwards
+    try:
+        with TinyDB(TMP_DB_PATH) as tmp_index_db:
+            for project in all_projects:
+                update_version_index_for_project(upload_folder_path, tmp_index_db, project.name)
+                update_file_index_for_project(upload_folder_path, tmp_index_db, project.name)
+    except Exception as e:
+        if TMP_DB_PATH.exists():
+            TMP_DB_PATH.unlink()
+
+        raise e
```

Comment:
Just replace this by a `finally` then you don't need to raise `e` agiain.

```suggestion
    finally:
        if TMP_DB_PATH.exists():
            TMP_DB_PATH.unlink()
```